File: go/bob/bob.go
// This is a "stub" file.  It's a little start on your solution.
// It's not a complete solution though; you have to write some code.

// Package bob should have a package comment that summarizes what it's about.
// https://golang.org/doc/effective_go.html#commentary
package bob

import (
	"strings"
	"unicode"
)

const (
	Whatever = "Whatever."
	Sure     = "Sure."
	Chill    = "Whoa, chill out!"
	CalmDown = "Calm down, I know what I'm doing!"
	Fine     = "Fine. Be that way!"
)

type Remark struct {
	phrase string
}

func (r Remark) isSilence() bool {
	return r.phrase == ""
}

func (r Remark) isShouting() bool {
	hasLetters := strings.IndexFunc(r.phrase, unicode.IsLetter) >= 0
	isUpcased := strings.ToUpper(r.phrase) == r.phrase

	return hasLetters && isUpcased
}

func (r Remark) isQuestion() bool {
	return strings.HasSuffix(r.phrase, "?")
}

func (r Remark) isExasperated() bool {
	return r.isShouting() && r.isQuestion()
}

func newRemark(remark string) Remark {
	return Remark{
		strings.TrimSpace(remark),
	}
}

// Hey should have a comment documenting it.
func Hey(remark string) string {
	r := newRemark(remark)
	switch {
	case r.isSilence():
		return Fine
	case r.isExasperated():
		return CalmDown
	case r.isShouting():
		return Chill
	case r.isQuestion():
		return Sure
	default:
		return Whatever
	}
}

File: zig/bob/bob.zig
const std = @import("std");
const allocator = std.heap.page_allocator;

const Reply = enum {
    Whatever,
    Sure,
    Chill,
    CalmDown,
    Fine,
};

fn isSilent(s: []const u8) bool {
    return s.len == 0;
}

fn isShouting(s: []const u8) bool {
    const s_upper = std.ascii.allocUpperString(allocator, s) catch unreachable;
    defer allocator.free(s_upper);
    const all_upper = std.mem.eql(u8, s_upper, s);

    var has_letters = false;
    for (s) |c| {
        if (std.ascii.isAlphabetic(c)) has_letters = true;
    }

    return all_upper and has_letters;
}

fn isQuestion(s: []const u8) bool {
    return s[s.len - 1] == '?';
}

fn parseResponse(s: []const u8) Reply {
    if (isSilent(s)) return Reply.Fine; // silent
    if (isShouting(s) and isQuestion(s)) return Reply.CalmDown; // exasperated
    if (isShouting(s)) return Reply.Chill; // shouting
    if (isQuestion(s)) return Reply.Sure; // question
    return Reply.Whatever;
}

pub fn response(s: []const u8) []const u8 {
    const trimmed = std.mem.trim(u8, s, " \t\r\n");
    const res = parseResponse(trimmed);

    switch (res) {
        .Sure => return "Sure.",
        .Chill => return "Whoa, chill out!",
        .CalmDown => return "Calm down, I know what I'm doing!",
        .Fine => return "Fine. Be that way!",
        .Whatever => return "Whatever.",
    }
}
